### PR TITLE
feat(SPV-730) TOTP and Contacts support

### PR DESCRIPTION
**Pull request by Nazarii-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: package.json
Comment (chris-4chain) [inline package.json:77]: We should check whether or not we want to rely on this lib. (its only single file src) https://github.com/emn178/hi-base32/blob/master/src/base32.js

Maybe we could embed our own implementation - it doesn't seem challenging

File: src/xpriv/xpriv.ts
Comment (chris-4chain) [inline src/xpriv/xpriv.ts:89]: Shouldn't we use the `ts-sdk` for that?

File: package.json
Comment (Nazarii-4chain) [reply package.json:73]: Removed packages

File: package.json
Comment (Nazarii-4chain) [reply package.json:78]: It does. And I found a solution that might help

File: package.json
Comment (Nazarii-4chain) [reply package.json:77]: Removed this package

File: src/xpriv/xpriv.ts
Comment (Nazarii-4chain) [reply src/xpriv/xpriv.ts:89]: Removed

**Review by dorzepowski — CHANGES_REQUESTED** [review]
Leaving Request Changes to prevent merge, because it is adding dependency to Buffer which is a problem in mobile app.
We don't want to merge it before we resolve the problem with Buffer in current js-client version.

File: src/utils/totp.ts
Comment (chris-4chain) [reply src/utils/totp.ts:96]: As I discussed it with mobile developer, it doesn't solve the problem.

**Review by dorzepowski — APPROVED** [review]

File: package.json
Comment (chris-4chain) [inline package.json:1]: We need to bump version field to the next value.

**Review by chris-4chain — CHANGES_REQUESTED** [review]

File: src/utils/totp.ts
Comment (chris-4chain) [inline src/utils/totp.ts:1]: This TOTP functionality should be available through the client - as it is implemented in `go-client`.
I'm suggesting, wrapping these functions within some "class" and instantiating this class on client creation. 

https://github.com/bitcoin-sv/spv-wallet-go-client/blob/94136a78cd3855649226980c256c1559df844e62/totp.go#L39

File: src/utils/totp.ts
Comment (chris-4chain) [inline src/utils/totp.ts:31]: I think, it is unsafe to do it this way (`totp.options = ...`)
Because of the nature of the `otplib`, we modifies here "global" options witch causes several issues:
- if a "user" of our `js-client` use the `totp` outside of "our" code - he will share the same options with js-client which may lead to undefined behaviour
- As creators of this lib say: "use epoch witch caution". Here we globally set the time base for all... I think we don't have to set it at all, because it set under-the-hood here: https://github.com/yeojz/otplib/blob/147e67637377196e3c59cf93b12ac325aca76fac/packages/otplib-core/src/totp.ts#L149
image

I can see following solutions:
- use otplib/core and create an instance of TOTP every time it is needed (with specialized options)
- use `clone()` before setting the options, like that:
```typescript
let secret = directedSecret(sharedSecret, contact.paymail);

  const totpInstance = totp.clone()
  totpInstance.options = {digits: digits, step: period}
  totpInstance.generate(secret)
```

**Review by chris-4chain — CHANGES_REQUESTED** [review]

File: src/utils/totp.ts
Comment (chris-4chain) [reply src/utils/totp.ts:1]: Apparently, there is (for now) no benefit of wrapping these functions in a class. But, I've made following methods for the client: `GenerateTotpForContact` & `ValidateTotpForContact`.

File: src/totp/totp.ts
Comment (wregulski) [inline src/totp/totp.ts:113]: It would be great to create an issue in "Discussion" type:
image
for them to consider and add this feature to the library :D 

// I don't like hanging TODOs :D

Comment (chris-4chain) [reply src/totp/totp.ts:113]: What I was planned is to show `ts-sdk` team the whole `totp` directory and discuss what to move to their package.
However, I'd prefer to do it after our internal rewiews & tests - so basically, after merging-  because I would not like the discussion to block this PR.

File: src/client.ts
Comment (yarex-4chain) [inline src/client.ts:749]: Are you sure this function call is not asynchronous like before?

Comment (yarex-4chain) [reply src/client.ts:749]: ahh, ok it returns a Promise, so it doesn't need to await

File: src/totp/converters.ts
Comment (yarex-4chain) [inline src/totp/converters.ts:5]: I would expect some comments or putting these numbers into constants for the logic to be more explanitory

File: src/totp/converters.ts
Comment (yarex-4chain) [inline src/totp/converters.ts:15]: Same thing here - why 15.5 for example?

File: src/totp/sha1hmac.ts
Comment (yarex-4chain) [inline src/totp/sha1hmac.ts:26]: Consider 0x36 as a constant

File: src/totp/sha1hmac.ts
Comment (yarex-4chain) [inline src/totp/sha1hmac.ts:32]: Consider 0x6a as a constant

File: src/totp/pike-totp.ts
Comment (yarex-4chain) [inline src/totp/pike-totp.ts:75]: What is m/0/0/0 ?

File: src/totp/totp.ts
Comment (yarex-4chain) [inline src/totp/totp.ts:104]: Consider moving 8 and 0x7fffffff to constants

**Review by yarex-4chain — COMMENTED** [review]
The code looks good, but I left some comments to consider about improving the code's readability

File: src/totp/converters.ts
Comment (chris-4chain) [reply src/totp/converters.ts:15]: I'll implement it differently to make it more readable.

File: src/totp/sha1hmac.ts
Comment (chris-4chain) [reply src/totp/sha1hmac.ts:26]: What I did for `SHA1HMAC` is: I've made an exact copy of the code from `ts-sdk` [SHA256HMAC](https://github.com/bitcoin-sv/ts-sdk/blob/8212c9d1cfce772d26de5ae8eff195ebc85c4cc4/src/primitives/Hash.ts#L1108-L1190) and adjust it to different `SHA` algorithm. 

I suppose that ts-sdk team will embed this class into the SDK so this file will be deleted. 
So, for now I'd prefer not to "improve" this code if it works.

File: src/totp/sha1hmac.ts
Comment (chris-4chain) [reply src/totp/sha1hmac.ts:32]: Same consideration as above ☝️

File: src/totp/pike-totp.ts
Comment (chris-4chain) [reply src/totp/pike-totp.ts:75]: Good you mention that 👍  I'll add the same comment which is made in `go-client` for the equivalent code. 

This is the derivation path for `PKI` - it's related to HD Keys.
I don't think it has to be explained here in the code - because it's more like "domain knowledge"

File: src/client.ts
Comment (chris-4chain) [reply src/client.ts]: ℹ️ Added

File: src/totp/converters.ts
Comment (chris-4chain) [reply src/totp/converters.ts]: ℹ️ I've added a comment and made a local variable`hexPair`.

File: src/totp/converters.ts
Comment (chris-4chain) [reply src/totp/converters.ts]: ℹ️ My previous implementation was quite smart 🙈  but unreadable at the same time. 
I've made an equivalent implementation which should work the same, and be quite self-explanatory.

File: src/totp/totp.ts
Comment (chris-4chain) [reply src/totp/totp.ts]: This the crucial part of the whole algorithm. 
I've put some effort to make it more readable - check it out. 
I've added comments and constants.

**Review by yarex-4chain — APPROVED** [review]

**Review by wregulski — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]

**Pull request merged**
